### PR TITLE
Port samples-js-shoppingcart to TypeScript

### PR DESCRIPTION
**Pull request by vkorenev**
Port of [samples-ts-shoppingcart](https://github.com/cloudstateio/samples-ts-shoppingcart) to TypeScript.

Changes to build process:
- build steps are moved into Docker in order to make build process less dependent on user setup. 
- a separate build stage was added in order to make the final image smaller.

Remaining issues:
- definitions generated by protobufjs cannot be used directly.
- type definition for Cloudstate is needed.
- `test.js` needs to be ported to TypeScript as well.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+```
+
+Add these routes by performing
+```
+kubectl apply -f routes.yaml -n <project-name>
+```
+
+Open a web browser and navigate to:
+
+`https://<project-name>.us-east1.apps.lbcs.io/pages/index.html`
```
Comment (marcellanz) [inline README.md:171]: Is this location meant to be in the OSS sample documentation?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+NAME                  AGE
+shopping-store   21m
+# verify stateful services
+$ kubectl -n <project-name>  get statefulservices
+NAME            AGE    REPLICAS   STATUS
+shopping-cart   7m     1          Running
+frontend        4m     1          Running
+```
+
+`https://<project-name>.us-east1.apps.lbcs.dev/pages/index.html`
```
Comment (marcellanz) [inline README.md:32]: Is this location meant to be in the OSS sample documentation? It seems to be here without context.

File: shopping-cart/src/shoppingcart.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    // And return the new state.
+    console.log("return state")
+    return cart
+}
+
+/**
+ * Handler for item removed events.
+ */
+function itemRemoved(removed: domain.ItemRemoved, cart: domain.Cart): domain.Cart {
+    // Filter the removed item from the items by product id.
```
Comment (marcellanz) [inline shopping-cart/src/shoppingcart.ts:163]: there is a log line for every event above, but not this one.

File: README.md [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+## Building and deploying the Sample application
+
+### Frontend Service
+
+The frontend service is a front end web application written in typescript. It is backed by a `stateless` service that will serve the compiled javacript, html and images. This service makes `grpc-web` calls directly to the other services to get the data that it needs.
```
Comment (marcellanz) [inline README.md]: typescript => TypeScript?

Comment (marcellanz) [inline README.md]: javacript => javascript => JavaScript?

File: README.md [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+
+#### Getting container image ready
+
+You can use pre-built `lightbend-docker-registry.bintray.io/cloudstate-samples/frontend:latest` container image available at Lightbend Cloudstate samples repository.
```
Comment (marcellanz) [inline README.md]: You can use _the_ pre-built

File: README.md [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+Alternatively, you can clone [cloudstateio/samples-ui-shoppingcart](https://github.com/cloudstateio/samples-ui-shoppingcart) repository and follow the instructions there to build an image and deploy it to your own container repository.
```
Comment (marcellanz) [inline README.md]: you can clone => you can clone _the_

Comment (marcellanz) [reply README.md]: our own container repository => our own container _image_ repository

File: README.md [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+Check that the services are running
+```
+$ kubectl get statefulservices -n <project-name>
+NAME            AGE    REPLICAS   STATUS
+shopping-cart   6m     1          Running
+frontend        3m     1          Running
+```
+
+To redeploy a new image to the cluster you must delete and then redeploy using the yaml file.  
+For example if we updated the shopping-cart docker image we would do the following.
```
Comment (marcellanz) [inline README.md]: For example => For example, if we updated the shopping-cart docker image, we would do

File: README.md [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+```
+$ kubectl delete statefulservice shopping-cart -n <project-name>
```
Comment (marcellanz) [inline README.md:133]: do you have to delete here? wouldn't it be just a `kubectl apply` for an update?

File: README.md [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+statefulservice.cloudstate.io "shopping-cart" deleted
+$ kubectl apply -f js-shopping-cart.yaml -n <project-name>    
+statefulservice.cloudstate.io/shopping-cart created
+```
+
+## Routes
+The last thing that is required is to provide the public routes needed for both the front end and grpc-web calls.  These exist in the `routes.yaml` file.
```
Comment (marcellanz) [inline README.md]: The last thing that is required, is to provide

File: README.md (hunk 4 above)
Comment (marcellanz) [inline README.md]: _frontend_ <=> _front end_ could be used consistently.

**Comment by vkorenev** [comment]
@marcellanz, thanks for your comments! I'll need to fix that in the JS and Java samples as well.

File: deploy/shopping-store.yaml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+# Deployment spec for a store. A store is a logical abstraction over a database
+# deployment, and captures how to deploy and connect to that database. Multiple
+# StatefulServices's would use a single store.
```
Comment (marcellanz) [inline deploy/shopping-store.yaml]: I think this is not a possessive form? And would be StatefulServices I think.

Comment (vkorenev) [reply deploy/shopping-store.yaml]: Yep, you're right! I copied that from the original JS sample, so it needs to be fixed there as well.

File: shopping-cart/package.json [hunk 11]
```diff
@@ -0,0 +1,58 @@
+{
+  "name": "shopping-cart",
+  "version": "0.1.0",
+  "description": "Sample CloudState event sourced shopping cart application",
```
Comment (marcellanz) [inline shopping-cart/package.json]: CloudState => Cloud**s**tate

Comment (vkorenev) [reply shopping-cart/package.json]: I've seen both variants in the docs. Is Cloudstate the correct spelling?

File: shopping-cart/test/test.js [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const path = require("path");
+const should = require('chai').should();
```
Comment (marcellanz) [inline shopping-cart/test/test.js]: `'chai'` => `"chai"` to be consistent

**Comment by marcellanz** [comment]
@vkorenev cool to see TypeScript used, thanks :) I added a few comments.

File: shopping-cart/package.json (hunk 11 above)
Comment (marcellanz) [reply shopping-cart/package.json]: I think so. LB changed the Marketing term late 2019 from CloudState to Cloudstate. You would find still classes and types named `CloudState` used in camelcase, but if we talk about a `Sample CloudState event sourced shopping cart application` I think we talk about Cloudstate. Probably not super important, but being consistent helps.

Comment (marcellanz) [reply shopping-cart/package.json]: the relevant PR https://github.com/cloudstateio/cloudstate/pull/131

**Pull request closed without merging**